File: my_av.py
#!/usr/bin/python3

def check_database(my_line):
    file = open("data/urls/domains_database", "rt")
    data = file.readlines()
    for domain in data:
        domain = domain.strip()
        if domain in my_line:
            return 1
    file.close()
    return 0

def check_malware(line):
    n = len(line)
    if line[n - 2] == 'e' and line[n - 3] == 'x' and line[n - 4] == 'e' and line[n - 5] == '.':
        return 1
    return 0

def check_digits(line, test):
    digits = 0
    len_dom = 0
    i = -1
    poz = 0
    i = line.find("http://")
    if i != -1:
        poz = 7
    else:
        i = line.find("https://")
        if i != -1:
            poz = 8 
    while line[poz] != '/' and poz < len(line) - 1:
        if line[poz] >= '0' and line[poz] <= '9':
            digits += 1
        poz += 1
        len_dom += 1

    if digits * 10 >= len_dom:
        return 1
    return 0

def check_www(line):
    poz = -1
    poz = line.find("http://")
    if poz == -1:
        poz = line.find("https://")
    if poz == -1:
        poz = 0
    if line[poz] == 'w' and line [poz + 1] == 'w' and line[poz + 2] == 'w':
        if line[poz + 3] != '.':
            return 1
    return 0

f = open("data/urls/urls.in", "rt")
g = open("urls-predictions.out", "wt")

lines = f.readlines()
test = 0
for line in lines:
    test += 1
    ok1 = check_database(line)
    ok2 = check_malware(line)
    ok3 = check_digits(line, test)
    ok4 = check_www(line)
    if ok1 == 1 or ok2 == 1 or ok3 == 1:
        g.write("1\n")
    elif ok4 == 1:
        g.write("1\n")
    else:
        g.write("0\n")


f.close()
g.close()

def valid_flow_time(line):
    poz = -1
    poz = line.find("0 days")
    if poz != -1:
        for i in range(poz, poz + 8):
            if line[i] != ':' and line[i] != '0':
                return 0
    return 1

def valid_flow_pkts(line):
    n = len(line)
    if line[n - 1] == '0' and line[n - 2] == '.' and line[n - 3] == '0' and line[n - 4] == ',':
        return 1
    return 0

fin = open("data/traffic/traffic.in", "rt")
fout = open("traffic-predictions.out", "wt")

test = 0
lines = fin.readline()
lines = fin.readlines()
for line in lines:
    line = line.strip()
    ok1 = valid_flow_time(line)
    ok2 = valid_flow_pkts(line)
    if ok1 == 1:
        fout.write("0\n")
    elif ok2:
        fout.write("0\n")
    else: # ok1 = 0 si ok2 = 0
        fout.write("1\n")

fin.close()
fout.close()

File: my_av.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#define NMAX 10000

// malware sau phishing
int check_database(char *s)
{
	FILE *data = fopen("data/urls/domains_database", "rt");
	while (!feof(data)) {
		char line[NMAX];
		fscanf(data, "%s", line);
		if (strstr(s, line))
			return 1; // is malicious from database
	}
	fclose(data);
	return 0; // cannot be found int the database
}

// checks if the URL is executable
int check_malware(char *s)
{
	int len = strlen(s);
	if (s[len - 1] == 'e' && s[len - 2] == 'x' && s[len - 3] == 'e' &&
		s[len - 4] == '.')
		return 1;
	if (strstr(s, ".exe"))
		return 1; // is malware
	// return 0 if it's not malware
	return 0;
}

// checks if the domain contains more than 10% digits
int check_digits(char *s)
{
	int len = strlen(s), digits = 0, len_dom = 0;
	int i = 0;
	//checks if the domain starts either with http or https or neither
	char *p = strstr(s, "http://");
	if (p) {
		i = 7;
	} else {
		p = strstr(s, "https://");
		if (p)
			i = 8;
	}
	while (s[i] != '/' && i < len) {
		if (s[i] >= '0' && s[i] <= '9')
			digits++;
		i++;
		len_dom++;
	}
	if (digits * 10 >= len_dom)
		return 1; // is malware or phishing
	return 0;
}

// checks if contains www in a normal format
int check_www(char *s)
{
	int poz = 0;
	char *p = strstr(s, "http://");
	if (p)
		poz = 7;
	p = strstr(s, "https://");
	if (p)
		poz = 8;
	if (s[poz] == 'w' && s[poz + 1] == 'w' && s[poz + 2] == 'w') {
		if (s[poz + 3] != '.')
			return 1;
	}
	return 0;
}

//checks if duration is longer than 1 sec
int valid_flow_time(char *s)
{
	char *p = strstr(s, "0 days");
	if (p) {
		for (int i = 7; i <= 14; i++)
			if (p[i] != ':' && p[i] != '0')
				return 0;
	}
	return 1; // e 0 sec
}

// checks if the line end with .0.0, which means it's valid
int valid_flow_pkts(char *s)
{
	int len = strlen(s);
	if (s[len - 1] == '0' && s[len - 2] == '.' && s[len - 3] == '0' &&
		s[len - 4] == ',')
		return 1; // e ,0.0
	return 0;
}

int main(void)
{
	// TASK 1
	FILE *f = fopen("data/urls/urls.in", "rt");
	FILE *g = fopen("urls-predictions.out", "wt");

	char *s = (char *)malloc(NMAX * sizeof(char));
	int nr = 0;
	while (!feof(f)) {
		fscanf(f, "%s\n", s);
		nr++;
		int ok1 = check_database(s);
		int ok2 = check_malware(s);
		int ok3 = check_digits(s);
		if (ok1 || ok2 || ok3) {
			fprintf(g, "1\n");
		} else {
			if (check_www(s))
				fprintf(g, "1\n");
			else
				fprintf(g, "0\n");
		}
	}
	free(s);
	fclose(f);
	fclose(g);

	// TASK 2

	FILE *in = fopen("data/traffic/traffic.in", "rt");
	FILE *out = fopen("traffic-predictions.out", "wt");

	char *sir = (char *)malloc(NMAX * sizeof(char));
	fgets(sir, NMAX, in);
	while (fgets(sir, NMAX, in)) {
		if (strchr(sir, '\n'))
			sir[strlen(sir) - 1] = '\0';
		if (valid_flow_time(sir)) {
			fprintf(out, "0\n");
		} else {
			if (valid_flow_pkts(sir))
				fprintf(out, "0\n");
			else
				fprintf(out, "1\n");
		}
	}
	free(sir);
	fclose(in);
	fclose(out);

	return 0;
}
